### PR TITLE
Add monthly variations for recurrence rules

### PR DESCRIPTION
**Pull request by allenporter**
## Proposed change

Add monthly variations for recurrence rules which includes things like:
- On the 20st day of the month (default)
- On the 3rd Wednesday
- On the last Tuesday

This requires passing in the start date to the recurrence rule editor which is used to determine the day of the week and rules that should apply. For Weekly recurrence rules, this also is used to  reset the weekday value (e.g. it automatically selects and updates the correct day of week)

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue:
- This PR is related to issue or discussion: Passing date to recurrence rule editor is a pre-req for solving issue #84042
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/panels/calendar/ha-recurrence-rule-editor.ts [hunk 1]
```diff
@@ -129,7 +152,30 @@ export class RecurrenceRuleEditor extends LitElement {
   }
 
   renderMonthly() {
-    return this.renderInterval();
+    return html`
+      ${this.renderInterval()}
+      ${this._monthlyRepeatItems.length > 0
+        ? html`<ha-select
+            id="monthly"
+            label="Repeat Monthly"
```
Comment (allenporter) [inline src/panels/calendar/ha-recurrence-rule-editor.ts:160]: I just realized there are many labels in this file that are not yet using translations, so i'll update all of them in a followup pass.

**Comment by allenporter** [comment]
One issue I need help with is that when changing the start date or interval, the selected item in drop down list is not invalidated:

Screen Shot 2022-12-21 at 9 35 31 AM

The items in the list are correct but not the selected item.  To fix this, do i need to grab this dom element with a selector and invalidate it explicitly somehow?

File: src/panels/calendar/ha-recurrence-rule-editor.ts [hunk 2]
```diff
@@ -129,7 +152,30 @@ export class RecurrenceRuleEditor extends LitElement {
...
+            @selected=${this._onMonthlyDetailSelected}
+            @closed=${stopPropagation}
+            fixedMenuPosition
+            naturalMenuWidth
+          >
+            ${this._monthlyRepeatItems!.map(
+              (item) => html`
+                <ha-list-item
+                  .value=${item.value}
+                  .selected=${item.byday === this._monthlyRepeat}
```
Comment (bramkragten) [inline src/panels/calendar/ha-recurrence-rule-editor.ts:170]:
```suggestion
```

**Review by bramkragten — DISMISSED** [review]

**Review by allenporter — COMMENTED** [review]
Have a question about the BYDAY addition...

File: src/panels/calendar/recurrence.ts [hunk 3]
```diff
@@ -138,3 +166,127 @@ export function ruleByWeekDay(
...
+      label: getLabel(`BYMONTHDAY=${getDate(dtstart)}`)!,
+    },
+    // Additional optional rules based on the week of month e.g. 2nd sunday of month
+    ...getWeekydaysForMonth(dtstart).map((item) => ({
+      value: `BYDAY=${item.toString()}`,
+      byday: item,
+      label: getLabel(`BYDAY=${item.toString()}`)!,
+    })),
+  ];
+  if (isLastDayOfMonth(dtstart)) {
```
Comment (allenporter) [inline src/panels/calendar/recurrence.ts:214]: Is this handled by `getWeekydaysForMonth` above when it does the `isLastWeekday` check?

File: src/panels/calendar/ha-recurrence-rule-editor.ts [hunk 4]
```diff
@@ -61,12 +80,46 @@ export class RecurrenceRuleEditor extends LitElement {
...
-    if (!changedProps.has("value") || this._computedRRule === this.value) {
+    if (changedProps.has("dtstart") || changedProps.has("_interval")) {
+      this._monthlyRepeatItems = this.dtstart
+        ? getMonthlyRepeatItems(this.hass, this._interval, this.dtstart)
+        : [];
+      this._computeWeekday();
+      const selectElement = this._monthlyRepeatSelect;
+      if (selectElement) {
+        const oldSelected = selectElement.index;
+        // this._monthlyRepeat = undefined;
```
Comment (allenporter) [inline src/panels/calendar/ha-recurrence-rule-editor.ts:91]: Not needed?

File: src/panels/calendar/recurrence.ts (hunk 3 above)
Comment (bramkragten) [reply src/panels/calendar/recurrence.ts:214]: I don't understand?

Comment (bramkragten) [reply src/panels/calendar/recurrence.ts:214]: That is the last weekday of the month, as in last friday of the month.

This is the last day of the month, so the 28, 29, 30 or 31.

**Pull request merged**